### PR TITLE
EFI: support for Lutan EFI system

### PR DESCRIPTION
**Pull request by tridge**
EFI system from this motor:
https://www.austars-model.com/dle60cc-efi-uav-engine-for-fixed-wing-drone-alternator-12v80w-ros-version-gst-inc-for-local-vip_g21812.html?s_id=4&fbclid=IwAR1XIRTI6cdM9UbjM6orE3bFIi-zsrqXwI-dKfdOmkWbuXWOaaLxZCOWIYQ

tested by Kyle Gardner

### CODE REVIEW

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+AP_EFI_Serial_Lutan::AP_EFI_Serial_Lutan(AP_EFI &_frontend):
+    AP_EFI_Backend(_frontend)
+{
+    port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_EFI, 0);
+}
+
+
+void AP_EFI_Serial_Lutan::update()
+{
+    if (!port) {
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:35]:
```suggestion
    if (port == nullptr) {
```

Your rule, not mine.

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:39]:
```suggestion
    const uint32_t now = AP_HAL::millis();
```

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        pkt_nbytes = 0;
+    }
+    const uint32_t n = MIN(sizeof(pkt), port->available());
+    if (n > 0) {
+        last_recv_ms = now;
+    }
+    if (n + pkt_nbytes > sizeof(pkt)) {
+        pkt_nbytes = 0;
+    }
+    pkt_nbytes += port->read(&pkt[pkt_nbytes], n);
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:50]: `read()` can return a negative value.  That could result in a bad memory access.

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    if (n == 0 || now - last_request_ms > 200) {
+        last_request_ms = now;
+        port->discard_input();
+        send_request();
+    }
+}
+
+void AP_EFI_Serial_Lutan::send_request(void)
+{
+    const uint8_t d[] = { 0, 1, 0x41 };
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:84]:
```suggestion
    static const uint8_t d[] { 0, 1, 0x41 };
```

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    const uint32_t crc = ~crc_crc32(~0U, &d[2], sizeof(d)-2);
+    const uint32_t crc2 = htobe32(crc);
+    port->write(d, sizeof(d));
+    port->write((const uint8_t *)&crc2, sizeof(crc2));
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:88]: Presumably we could pre-calculate this entire packet - but there might be different requests in the future so probably worth keeping the flexibility.

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            int16_t veCurr1; // %, scale 0.1
+            int16_t veCurr2; // %, scale 0.1
+            int16_t iacstep;
+            int16_t idleDC; // scale 0.392
+            int16_t coldAdvDeg; // deg, scale 0.1
+            int16_t TPSdot; // %/s, scale 0.1
+            int16_t MAPdot; // kPa/s
+            int16_t dwell; // ms, scale 0.0666
+        } data;
+        uint8_t pkt[400];
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.h:78]: 400 seems a little over the top?  That's a chunk of memory to never use.

File: libraries/AP_Math/AP_Math.cpp [hunk 7]
```diff
@@ -473,3 +473,9 @@ float fixedwing_turn_rate(float bank_angle_deg, float airspeed)
     bank_angle_deg = constrain_float(bank_angle_deg, -80, 80);
     return degrees(GRAVITY_MSS*tanf(radians(bank_angle_deg))/MAX(airspeed,1));
 }
+
+// convert degrees farenheight to Kelvin
+float degF_to_Kelvin(float temp_f)
+{
+    return (temp_f + 459.67) * 0.55556;
```
Comment (peterbarker) [inline libraries/AP_Math/AP_Math.cpp:480]: Surely in-lining this is fewer bytes than making a call?

File: libraries/AP_SerialManager/AP_SerialManager.cpp [hunk 8]
```diff
@@ -547,8 +547,8 @@ void AP_SerialManager::init()
                     break;
 #endif
                     
-                case SerialProtocol_EFI_MS:
-                    state[i].baud = AP_SERIALMANAGER_EFI_MS_BAUD;   // update baud param in case user looks at it
+                case SerialProtocol_EFI:
+                    state[i].baud.set_default(AP_SERIALMANAGER_EFI_MS_BAUD);
```
Comment (peterbarker) [inline libraries/AP_SerialManager/AP_SerialManager.cpp:551]: Still referencing `_MS` here.  I guess it's the default for that particular device, though....

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.h (hunk 6 above)
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_Lutan.h:78]: it does use it. The packets are quite large and variable, we just get all the data we need from the front.

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:50]: only if available() is broken. If it is then we have bigger problems

File: libraries/AP_SerialManager/AP_SerialManager.cpp (hunk 8 above)
Comment (tridge) [reply libraries/AP_SerialManager/AP_SerialManager.cpp:551]: indeed

File: libraries/AP_Math/AP_Math.cpp (hunk 7 above)
Comment (tridge) [reply libraries/AP_Math/AP_Math.cpp:480]: not tested, but i'd guess not inlining is better

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp]: done

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp]: done

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+#include <AP_HAL/AP_HAL.h>
+#include "AP_EFI_Serial_Lutan.h"
+#include <AP_HAL/utility/sparse-endian.h>
+#include <stdio.h>
+
+#if HAL_EFI_ENABLED
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:21]: This should really be `AP_EFI_SERIAL_LUTAN_ENABLED`

Nobody running this is really going to care, but it is a pattern we should spread as widely as we can.

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            const uint32_t crc2 = ~crc_crc32(~0U, &pkt[2], length);
+            if (crc == crc2) {
+                // valid data
+                internal_state.last_updated_ms = now;
+                internal_state.spark_dwell_time_ms = int16_t(be16toh(data.dwell))*0.1;
+                internal_state.cylinder_status[0].injection_time_ms = be16toh(data.pulseWidth1)*0.00666;
+                internal_state.engine_speed_rpm = be16toh(data.rpm);
+                internal_state.atmospheric_pressure_kpa = int16_t(be16toh(data.barometer))*0.1;
+                internal_state.intake_manifold_pressure_kpa = int16_t(be16toh(data.map))*0.1;
+                internal_state.intake_manifold_temperature = degF_to_Kelvin(int16_t(be16toh(data.mat))*0.1);
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:69]:
```suggestion
                internal_state.intake_manifold_temperature = F_TO_KELVIN(int16_t(be16toh(data.mat))*0.1);
```

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                internal_state.coolant_temperature = degF_to_Kelvin(int16_t(be16toh(data.coolant))*0.1);
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp:70]:
```suggestion
                internal_state.coolant_temperature = F_TO_KELVIN(int16_t(be16toh(data.coolant))*0.1);
```

File: libraries/AP_EFI/AP_EFI_Serial_Lutan.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+ */
+/*
+  support for Lutan serial EFI
+ */
+#pragma once
+
+#include "AP_EFI.h"
+#include "AP_EFI_Backend.h"
+
+class AP_EFI_Serial_Lutan: public AP_EFI_Backend {
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_Lutan.h:23]: This definition should be gated on `AP_EFI_ENABLED` (or preferable `AP_EFI_SERIAL_LUTAN_ENABLED`...)

File: libraries/AP_EFI/AP_EFI_Serial_MS.cpp [hunk 13]
```diff
@@ -110,12 +110,12 @@ bool AP_EFI_Serial_MS::read_incoming_realtime_data()
             case MAT_MSB:
                 temp_float = (float)((data << 8) + read_byte_CRC32())/10.0f;
                 offset++;
-                internal_state.intake_manifold_temperature = f_to_k(temp_float);
+                internal_state.intake_manifold_temperature = degF_to_Kelvin(temp_float);
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_Serial_MS.cpp:113]:
```suggestion
                internal_state.intake_manifold_temperature = F_TO_KELVIN(temp_float);
```

File: libraries/AP_Math/AP_Math.cpp (hunk 7 above)
Comment (peterbarker) [inline libraries/AP_Math/AP_Math.cpp:480]:
```suggestion
```
replace with F_TO_KELVIN

**Review by peterbarker — APPROVED** [review]
LGETM.

**Pull request merged**